docs/design/yarn-v1.md

```diff
@@ -0,0 +1,135 @@
+and refactored/improved.
+
+#### Key Points
+ - The yarn.lock file can be parsed with [pyarn](https://github.com/containerbuildsystem/pyarn) like in cachito
+ - Workspaces will need to be [processed separately](https://github.com/containerbuildsystem/cachito/blob/63b8ec0ea615d114ccfa0d08dc0bec49e60e6a75/cachito/workers/pkg_managers/yarn.py#L95) from package.json since they do not appear in the yarn.lock file
+ - Dev dependencies can be [determined similiarly](https://github.com/containerbuildsystem/cachito/blob/63b8ec0ea615d114ccfa0d08dc0bec49e60e6a75/cachito/workers/pkg_managers/yarn.py#L140) to cachito
+ - We will need to ensure that there are no collisions in the names of the archive files being added to the offline mirror
+
+
+### Avoiding Arbitrary Code Execution
+Yarn 1.x appears to honor the ```--ignore-scripts``` option for the install command. It can also be configured via environment variable.
+Used https://github.com/chmeliik/js-lifecycle-scripts/tree/yarn to verify.
+
+### Prefetch
+The offline mirror in the request output directory will be populated by running the following Yarn command in the project root:
+
+    yarn install --non-interactive --frozen-lockfile --disable-pnp --no-default-rc --ignore-engines
+
+Command: [yarn install](https://classic.yarnpkg.com/en/docs/cli/install)
+
+Options:
+ - non-interactive: Disable interactive prompts
+ - frozen-lockfile: Fail if yarn.lock updates are needed
+ - disable-pnp: Disable Plug'n'Play installation
+ - no-default-rc: Prevent Yarn from automatically detecting .yarnrc and .npmrc files
+ - ignore-engines: Ignore the engine node version during the prefetch
+
+ The behavior of Yarn can also be configured using select environment variables. The following ones should be set for the prefetch:
+
+    COREPACK_ENABLE_DOWNLOAD_PROMPT=0
+    COREPACK_ENABLE_PROJECT_SPEC=0 (If we decide to use the global default)
+    YARN_YARN_OFFLINE_MIRROR=<absolute path to the request output directory>
+    YARN_YARN_OFFLINE_MIRROR_PRUNING=false
+    YARN_IGNORE_PATH=true (Use the corepack yarn version, not any Yarn version specified by the user)
+    YARN_IGNORE_SCRIPTS=true
+
+### Build Configuration
+The user build can be configured to use the offline mirror by configuring the following environment variables:
+
+    YARN_YARN_OFFLINE_MIRROR=<request output directory>
+    YARN_YARN_OFFLINE_MIRROR_PRUNING=false
+
+## High-Level Procedure
+ - Since we're executing Yarn commands directly, process the request in an [isolated temporary directory](https://github.com/containerbuildsystem/cachi2/blob/6953607b6ef52fd3f0bef7059d2c926767b1022b/cachi2/core/resolver.py#L41) to avoid any potential changes to the repository
+ - Read the project files
+   - Read/Require package.json
+   - Read/Require yarn.lock
+     - Ensure yarn.lock is formatted for Yarn 1.x
```

Comment:
I've successfully convinced myself of both options so far, so I'm curious what you all think :sweat_smile:

Reply:
I'd lean towards `yarn`, feels more user-friendly

Reply:
+1 on `yarn` (or possibly `yarn-classic` , `yarn-berry`)
IIRC we plan to support yarn v4 in the future, so we would need three yarn types, right ?

Reply:
Not really, we support multiple versions of each package manager without asking the user to specify the exact version.

Yarn 1 vs. Yarn 2+ is an exception - they are so different that they have less in common than NPM and Yarn.

But from the users' point of view, it's still just `yarn`

Reply:
> But from the users' point of view, it's still just yarn

it is, but for us it's very different, so in this case I'd adhere to sticking to the official names --> `yarn-classic` because otherwise we'll have figure out ourselves which implementation to use, is that the intention here, do it dynamically or do we require the user to specify it? Because if it's supposed to be dynamic then I have to retract on the name preference as we'll have to cope with just `yarn`.

Reply:
Right, I noticed that bit somewhere in the document in the end.